sdk/src/compressed_account.rs

```diff
@@ -1,421 +1,152 @@
 use light_utils::hash_to_bn254_field_size_be;
+use solana_program::account_info::AccountInfo;
 
 use crate::{
-    address::{derive_address, NewAddressParamsPacked},
-    merkle_context::{
-        pack_merkle_context, MerkleContext, PackedAddressMerkleContext, PackedMerkleContext,
-        RemainingAccounts,
+    account_info::LightAccountInfo,
+    address::{
+        derive_address_from_params, derive_address_seed, unpack_new_address_params,
+        PackedNewAddressParams,
     },
-    program_merkle_context::unpack_address_merkle_context,
+    merkle_context::{pack_merkle_context, MerkleContext, PackedMerkleContext, RemainingAccounts},
 };
 
-pub trait LightAccounts: Sized {
-    fn try_light_accounts(
-        inputs: Vec<Vec<u8>>,
-        merkle_context: PackedMerkleContext,
-        merkle_tree_root_index: u16,
-        address_merkle_context: PackedAddressMerkleContext,
-        address_merkle_tree_root_index: u16,
-        remaining_accounts: &[AccountInfo],
-    ) -> Result<Self>;
-    fn new_address_params(&self) -> Vec<NewAddressParamsPacked>;
-    fn input_accounts(
-        &self,
-        remaining_accounts: &[AccountInfo],
-    ) -> Result<Vec<PackedCompressedAccountWithMerkleContext>>;
-    fn output_accounts(
-        &self,
-        remaining_accounts: &[AccountInfo],
-    ) -> Result<Vec<OutputCompressedAccountWithPackedContext>>;
+pub trait LightAccounts<'a>: Sized {
+    fn try_light_accounts(accounts: &'a [LightAccountInfo]) -> Result<Self>;
 }
 
+// TODO(vadorovsky): Implment `LightAccountLoader`.
+
 /// A wrapper which abstracts away the UTXO model.
-pub enum LightAccount<T>
+pub struct LightAccount<'info, T>
 where
-    T: AnchorDeserialize + AnchorSerialize + Clone + DataHasher + Discriminator,
+    T: AnchorDeserialize + AnchorSerialize + Clone + DataHasher + Default + Discriminator,
 {
-    Init(LightInitAccount<T>),
-    Mut(LightMutAccount<T>),
-    Close(LightCloseAccount<T>),
+    account_state: T,
+    account_info: &'info LightAccountInfo<'info>,
+    new_address_params: Option<PackedNewAddressParams>,
+    address: Option<[u8; 32]>,
 }
 
-impl<T> LightAccount<T>
+impl<'info, T> LightAccount<'info, T>
 where
     T: AnchorDeserialize + AnchorSerialize + Clone + DataHasher + Default + Discriminator,
 {
-    pub fn new_init(
-        merkle_context: &PackedMerkleContext,
-        address_merkle_context: &PackedAddressMerkleContext,
-        address_merkle_tree_root_index: u16,
-    ) -> Self {
-        Self::Init(LightInitAccount::new(
-            merkle_context,
-            address_merkle_context,
-            address_merkle_tree_root_index,
-        ))
+    pub fn new(account_info: &'info LightAccountInfo) -> Result<Self> {
+        let account_state = if account_info.input.is_some() {
+            if let Some(data) = account_info.data {
+                T::try_from_slice(data)?
+            } else {
+                T::default()
+            }
```

Comment:
Why should the input account be default if there is no data submitted?
Shouldn't it be `None` in this case?

No input data implies a new account, hence I can see that it makes sense to set default in line 46. 

The use cases I can think of when you wouldn't want to submit any data are:
1. if you want to send only partial data and can infer remaining data
2. if you want to send partial or no data and just send the data hash as a public input to verify a program custom application zkp

Reply:
I agree with what you say. What I am saying is if there is an input commitment the struct shouldn't be instantiated with default data because the input commitment is most likely not to default data.

These are the three conditional branches in your code see below:
1. no input commitment & no input data -> Default (makes sense, handled in line 46)
2. input commitment & data -> Deserialized struct (makes sense, see line 41)
3. input commitment & no data -> Default (line 43, feels confusing to me because there is an input commitment which committed to data other than default but we instantiate the compressed account with default data.)


case 1 and case 2 make sense, case 3 should throw an error imo


```
pub fn new(account_info: &'info LightAccountInfo) -> Result<Self> {
        let account_state = if account_info.input.is_some() {
            if let Some(data) = account_info.data {
                T::try_from_slice(data)? // case 2
            } else {
                // T::default() // case 3
                err!(NoDataProvided)
            }
        } else {
            T::default() // case 1
        };
        Ok(Self {
            account_state,
            account_info,
            new_address_params: None,
            address: None,
        })
    }
 ```

Reply:
Ah, gotcha, thant makes sense - thanks!